metronome/networking/src/io/iohk/metronome/networking/ConnectionHandler.scala

```diff
@@ -0,0 +1,215 @@
+package io.iohk.metronome.networking
+
+import cats.effect.{Concurrent, ContextShift, Resource}
+import cats.effect.concurrent.{Deferred, TryableDeferred}
+import io.iohk.metronome.networking.RemoteConnectionManager.{
+  MessageReceived,
+  UnexpectedConnectionError,
+  withCancelToken
+}
+import monix.catnap.ConcurrentQueue
+import monix.execution.atomic.AtomicInt
+import monix.tail.Iterant
+import cats.implicits._
+import cats.effect.implicits._
+import io.iohk.metronome.networking.ConnectionHandler.HandledConnection
+import io.iohk.metronome.networking.EncryptedConnectionProvider.ConnectionError
+
+import java.net.InetSocketAddress
+
+class ConnectionHandler[F[_]: Concurrent, K, M](
+    connectionQueue: ConcurrentQueue[F, HandledConnection[F, K, M]],
+    connectionsRegister: ConnectionsRegister[F, K, M],
+    messageQueue: ConcurrentQueue[F, MessageReceived[K, M]],
+    cancelToken: TryableDeferred[F, Unit],
+    connectionFinishCallback: HandledConnection[F, K, M] => F[Unit]
+) {
+
+  private val numberOfRunningConnections = AtomicInt(0)
+
+  private def closeAndDeregisterConnection(
+      handledConnection: HandledConnection[F, K, M]
+  ): F[Unit] = {
+    for {
+      _ <- Concurrent[F].delay(numberOfRunningConnections.decrement())
+      _ <- handledConnection.close()
+      _ <- connectionsRegister.deregisterConnection(handledConnection)
+    } yield ()
+  }
+
+  /** Registers connections and start handling incoming messages in background, in case connection is already handled
+    * it closes it
+    *
+    * @param possibleNewConnection, possible connection to handle
+    */
+  def registerIfAbsent(
+      possibleNewConnection: HandledConnection[F, K, M]
+  ): F[Unit] = {
+    connectionsRegister.registerIfAbsent(possibleNewConnection).flatMap {
+      case Some(_) =>
+        //TODO for now we are closing any new connections in case of conflict, we may investigate other strategies
```

Comment:
tbh I am not sure what is the best default here. My intuition is that kinda depends on what we want to optimise for i.e full cluster restart, node come back after failure etc. For example exchanging old connection probably would be better in case of failed node coming back and connecting back to cluster. 

My take here, it that keeping what we have is pretty good starting point which, and we can iterate on it. I can create task for it and take it after this one. wdyt ? I would prefer to implement connection replacing logic in separate pr either way as involves adding a little more book keeping to make it safe.

Reply:
It looks to me like the other option, keeping the new connection and closing the old, is the better strategy if we think about honest scenarios, which is that one of us knows that the connection is broken, for whatever reason, and the other one doesn't, because detection only happens when one of the parties tries to send a message, or, as you say, went down and is trying to come back. Keeping the old connection would more likely be futile, with no more messages coming. 

Keeping the new should allow quicker recoveries: in a federation of 10 nodes, if Node 10 goes down, Node 1 will not try to talk to it until Node 1 becomes a leader of a round, so it can take a while until it tries to send a message and realises that it can't, while Node 10 is banging on the door and getting it slammed in its face.

So I don't see what's the benefit of using the keep-old strategy as default rather than the keep-new. If you're not convinced, sure, let's make another ticket, just didn't want it to be left like this.

Reply:
It make sense for incoming connections, but on the other hand for out going ones the conflicts can happen only when we are calling the peer and in the same time we have received incoming call, in that case I would argue it is better to keep connection we already have and close newly established outgoing. Does both strategies, replacing in case of incoming and keeping old in case of outgoing, are compatible ? tbh I am not sure, and would need to think more on this.

I have created task to take just after this one - https://jira.iohk.io/browse/PM-3092, to focus only this problem.

Reply:
Thanks for creating the ticket.

Yeah it's difficult to think it through. In theory the `ConnectionHandler` keeps the message sending and receiving decoupled from the connections,  so Alice can send a request on connection 1, and receive a response on connection 2, it would go in the same queue in the end. 

So there are two scenarios: 1) where there are two simultaneous open connections, in which case both parties would have to close the same one (which is where I thought you could order them by the ephemeral IP/port combo for example) for messages not to get lost and 2) where one of them already knows the connection they used to have is broken, but the other one doesn't, and is trying to re-establish. 

For the 2nd scenario, accepting the new connection is better than keeping the old. For the 1st, I think both old and new strategies are equal, because Alice and Bob can see them the opposite way, depending on timing, so whichever they keep might end up closing both connections.

So I think the good strategy is perhaps to check how old a connection is: if it's just been opened, then we can assume we could be in scenario 1 and pick one based on a symmetric ordering function (e.g. IP+port); if it's been opened a while ago, then we can assume it's scenario 2 and keep the new one.

Reply:
so for case 
1), are your thinking about something along the lines `may ephemeral port is X and remote ephemeral is Y, X< Y therefore I chose X` ? I was thinking about something similar but with keys i.e define some key metric, lets say xor distance from key `0000....` and in case of conflict higher one wins.
and fort case
2) the problem is to define what does it mean `it's just been opened` so when to resolve conflict or when to replace, taking into account also the fact that ultimatly the side which assumes everything will at some point learn that it is really not fine. 

Another design point, is to just keep both connections i.e one incoming and one outgoing to the same peer, and maybe have one connection for high important traffic, and second one as heart beating or some less important messages. (from what I remember nodes in elastic search cluster keep several open connections for different messages types)

Thats why I opened separate ticket to focus on just this problem

Reply:
1) I didn't really think it through. 
Say we have these two connections to consider:
Connection 1: `123.456.789.012:65432 -> 234.567.890.123:40000`
Connection 2: `234.567.890.123:54321 -> 123.456.789.012:40000`
Assuming that the listener port is 40000, then as you say you can compare the ephemeral 54321 against 65432 or combine that with IPs as well in case there ports are not conclusive. I thought this is something which is available even without keys, just by comparing sockets, so it should be easier to implement. They can't all be the same, IPs and ports, so it should work.

2) I don't know, a configurable timeout, like the minimum backoff period? Say Alice "just" opened a connection to Bob, yet she sees that Bob is trying to connect to her. If less than the minimal backoff period has elapsed since the existing connection has been open, Alice could assume that they just initiated both around the same time, by accident. If she closes the connection and Bob still hasn't realised that he's got an incoming one from her, he'll back off and try again. If Bob is indeed trying again after the backoff period, then maybe something is indeed wrong with the connection she made earlier, and we can trust Bob that he's trying to do the right thing. 

I can't actually remember in the code whether Bob will backoff, connect, try to register and discard when he realises that that there's an incoming already. He'd also have to make sure that he keeps the second connection, if Alice picks it. Or, better, after the backoff period, he should check whether an incoming connection exists, before trying to establish an outgoing one. 

What you're saying about keeping multiple connections - that sounds an even bigger design space, so I'd consider that as yet another ticket.

Reply:
`1` would require little tweaj in scalanet as currently it does not expose local ephemeral port to the user i.e when we connect to remote host we do not known what ephemeral port we got assigned for our

`2`  sounds reasonable but again we would probably need experiment a little bit as every timing assumption can bite us back, as assumption :`If less than the minimal backoff period has elapsed since the existing connection has been open, Alice could assume that they just initiated both around the same time, by accident` may not hold if there was some additional delay. (garbage collector, some additional traffic, other connections is queue before we registred new one etc.)

about ` Or, better, after the backoff period, he should check whether an incoming connection exists, before trying to establish an outgoing one` this is as it is currently handled in code i.e before making outgoing connection we always check that we did not received incoming one from the same peer, so the only reason we can have conflict in case of outgoing connection is the fact that we received incoming while we were calling.

Sure I will deal with  https://jira.iohk.io/browse/PM-3092, and later  if necessary we can talk if having multiple connections to one peer is something we would like to have.

Reply:
> this is as it is currently handled in code 

Awesome :muscle: 

>  scalanet as currently it does not expose local ephemeral port

Thanks, forgot about that.